#**MongoDB with *pymongo* (Practice Case)**
<br>
<br>
<br>
Zumarsiyah Mahsyari<br>
Astra Data Scientist Bootcamp

---

<br>
MongoDB is a document-oriented NoSQL database used for high volume data storage. MongoDB is a database that came into light around the mid-2000s. It falls under the category of a NoSQL database. MongoDB stores data in JSON-like documents, which makes the database very flexible and scalable. Python needs a MongoDB driver to access the MongoDB database. In this practice case, we will use the MongoDB driver, which is PyMongo to manage unstructured databases. What we will do in this practice case this as follows:<br>

1.   Create a new collection with the name clean_movies that is the same as the movies collection in the sample_mflix database using the movies_intial collection (projecting_queries pymongo)
2.   Validate the collection that has been made with the following parameters:

*   All documents on clean_movies and movies are the same
*   Amount of documents on clean_movies and movies are the same
*   All fields in clean_movies are in the movies

Let's start it.


# Create a new collection

First, we have to download and install a MongoDB driver.

In [130]:
!pip install pymongo
!pip install pymongo[srv]
!pip install pprint
!pip install dnspython

After that, We have to get the connecting with databases instance on MongoDB Cluster using MongoClient.

In [68]:
### import MongoClient class from pymongo module
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

import datetime

#### initiate connection ###
client = MongoClient("mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority")

print(client)

MongoClient(host=['cluster0-shard-00-00-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-01-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-02-miqju.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


Now, we can find out what databases are available in the MongoDB Cluster with the following code.

In [73]:
### show all db in cluster ###
print(client.list_database_names())

['sample_mflix', 'admin', 'local']


We will use the 'sample_mflix' database and the 'movies_initial' collection.

In [74]:
db = client['sample_mflix']
col = db.movies_initial

### show all collection in used db ###
print(db.list_collection_names())

['clean_movies_Putrisqiana', 'clean_movies_tara', 'clean_movies_bunga', 'clean_movies_zumar', 'movies_initial', 'clean_movies_putrisqiana', 'clean_movies_firdaus', 'clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_faizah']


Let's take a look at one of the data that came in movies_initial.

In [75]:
movieinitial = col.find_one()
movieinitial

{'_id': ObjectId('5daffb1b986c4adedb754af9'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer': '',
 'year': 1894}

Because the MongoDB cluster that we are using does not have a 'movies' collection, we need to get a collection of 'movies' using another Pymongo cluster.

In [76]:
client2 = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority")
db2 = client2['sample_mflix']
col2 = db2.movies

### show all collection in used db ###
print(db2.list_collection_names())

['clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_devita', 'movies_initial', 'clean_movies_frans', 'clean_movies_tara', 'clean_movies_tunggal', 'clean_movies', 'users', 'clean_movies_nurrimah', 'movies_testtopik', 'movies_scratch', 'clean_movies_topik', 'movies', 'clean_movies_afifah', 'clean_movies_andreas', 'clean_movies_indra_1']


Let's take a look at one of the data that came in movies_initial too.

In [77]:
movie = col2.find_one()
movie

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
  'Thomas H. Ince',
  'Raymond B. West',
  'Walter Edwards',
  'David Hartford',
  'Jay Hunt',
  'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the women are against it, people are pressed into service. A count has constructed a submarine and gets the order to sink an ocean liner, that is also carrying - supposedly - ammunition for the enemy. The count refuses to fire the torpedos, and sinks the submarine. He survives, but in a limbo between death and life where he meets Jesus, who takes him over to preach peace. Naturally the king arrests him and sentences him to death for treason, but then Jesus shows him the real face of war.',
 'genres': ['Drama'],
 'imdb': {'id': 6517, 'rating': 6.3, 'votes': 162},
 '

The two collections above turned out to have differences. In the movie_initial collection, there are 3 columns about IMDb, namely imdbID, IMDb rating, and imdbVotes. While in collection movies, the three columns are an array in the IMDb column. Then, we will make the three IMDb columns in the movie_initial collection an array as like as in the 'movies' collection.

In [80]:
imdb = list(col2.find({}, {'_id':0, 'imdb':1}))
imdb_list=[]
print(len(imdb))
for tit in imdb:
  imdb_list.append(list(tit.values())[0])
  
imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

In [81]:
imdb = []
for dic in imdb_list:
  imdb.append(dic['id'])
imdb[:5]

[6517, 8133, 12512, 13025, 15498]

After that, let's create a 'clean_movies_zumar' collection from 'movie_initial' collection using projecting queries.

In [123]:
pipeline = [
  {
        '$project': {
            ### this code is to include the keys for a new collection on database ###
            ### 1 for include , 0 for exclude the keys ###
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },                                                      
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'released': { '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] },
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            
            'id':1,
            'title': 1,
            'year': 1,
            'cast': 1,
            'fullplot': 1,
            'lastupdated': 1,
            'plot': 1,
            'poster': 1,
            'type': 1,
            'metacritic': 1,
            
            ### here we split some keys into an array value, renaming as well ###
            'directors': {'$split': ["$director", ", "]},
            'cast': {'$split': ["$cast", ", "]},
            'writers': {'$split': ["$writer", ", "]},
            'genres': {'$split': ["$genre", ", "]},
            'countries': {'$split': ["$country", ", "]},
            'awards': {'$split': ["$awards", ", "]},
            'languages': "$language",
            
            ### adding embeded documents ###
            'num_mflix_comments':"",
            'rated':"",
            ### grouping the several keys into embeded documents by one keys ###
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            'tomatoes': {
                'lastUpdated':"",
                'viewer': {'numReviews':"", 'rating':""}
            }
        }
    },
    {
        "$match": {"imdb":{"$in":[dict(sorted(imdb_list[x].items(), key=lambda x:x[0])) for x in range(len(imdb_list))]}}
    },
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies_zumar"
    }
]

pprint.pprint(list(client.sample_mflix.movies_initial.aggregate(pipeline)))

[]


Let's take a look at one of the data that came in this collection.

In [124]:
colzum = db.clean_movies_zumar
cmz = colzum.find_one()
cmz

{'_id': ObjectId('5daffb1b986c4adedb754afa'),
 'awards': ['1 win.'],
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'languages': '',
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'metacritic': '',
 'num_mflix_comments': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'poster': '',
 'rated': '',
 'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
 'runtime': '1',
 'title': 'Blacksmith Scene',
 'tomatoes': {'l

# Validate the collection that has been made

##All documents on 'clean_movies' and 'movies' are the same

To check whether all documents on 'clean_movies' and 'movies' are the same, we can use the following code.

In [125]:
grup = list(db.clean_movies_zumar.aggregate([{"$sortByCount":"$imdb"}]))
grup2 = list(db2.movies.aggregate([{"$sortByCount":"$imdb"}]))

for i in grup2:
  if i not in grup:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


After we compare based on the 'imdb' field, there are document differences between the two collections.

In [126]:
list((colzum.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189 }})))

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': ['1 win.'],
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'languages': '',
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'metacritic': '',
  'num_mflix_comments': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'poster': '',
  'rated': '',
  'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
  'runtime': '1',
  'title': 'Blacksmith Scene'

In [127]:
list((col2.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189 }})))

[{'_id': ObjectId('5dafd0d3d75e5fa8b23dae21'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short', 'Long'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene II',
  'to

We can see that at the same imdb value, 'clean_movies_zumar' collection has a document, while 'movies' collection has 2 documents.

## The number of documents on 'clean_movies' and 'movies' are the same

Let's compare the number of documents in the 'clean_movies' collection and 'movies' collection.

In [93]:
len(list(col2.find()))

23541

In [86]:
len(list(colzum.find()))

23539

We can see that many documents in the two collections are different. 'clean_movies_collection' has 23,541 documents, while 'movies' collection has 23,549 documents.

## All fields in 'clean_movies' are in the 'movies'

Let's compare all the fields in the 'clean_movies' collection and 'movies' collection.

In [128]:
b = [list(doc.keys()) for doc in col2.find({})]
b = set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [129]:
c = [list(doc.keys()) for doc in colzum.find({})]
c = set([j for i in b for j in i])
c

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

We can see that the two collections above have the same fields, which means that all fields in 'clean_movies' collection are in the 'movies' collection.

# Conclusion

The conclusion that can be drawn in this practice case are:


1.   'clean_movies_zumar' collection has successfully created from 'movies' collection using projecting query.
2.   Based on the result of validation, 'clean_movies_zumar' collection and 'movies' collection have a different characteristics.

